### DISCUSSION TITLE
How obtain a delivered status mail by reference?

### DISCUSSION THREAD

**Discussion by diegovn07** [jstedfast/MailKit#1716] [General]
Hi, I am implementing a functionality that sends a list of emails and I need to check if any were not delivered. I have followed these instructions https://stackoverflow.com/questions/48118781/asp-net-mailkit-smtp-response and I already get the failed email when one is not delivered, but since it is an email list that is sent, how do you know which one it corresponds to? I could see that the delivery status notification emails include in the header a reference that matches the messageId, but when I try to search for emails with that reference I only get an empty list.

It is the first time I implement something similar, I have never worked with emails, I appreciate any help or correction

![messageid](https://github.com/jstedfast/MailKit/assets/66571905/ee2462df-ac70-4166-a516-43db37300005)
![References](https://github.com/jstedfast/MailKit/assets/66571905/db14a034-5bd4-4619-a9ec-954d467ec6a3)

**Comment by jstedfast** [comment]
The `References` header contains the `Message-ID` value of the message it references.

In other words, if you are trying to get the MimeMessage that is being referenced, you would search for the message that has a `Message-ID` header with a value of `"<" + messageId + ">"`:

```csharp
var email = await inbox.SearchAsync (SearchQuery.HeaderContains ("Message-ID", "<" + messageId + ">"));
```

This, of course, assumes that the referenced message exists in the INBOX folder (which it probably doesn't if you are trying to find the message that was sent because it'd probably be in a "Sent" folder).

Let me ask you this: why do you need the original message?

If I was writing an automated system that did this sort of thing, the only purpose for me would be to verify that this Delivery Status Notification email wasn't fraudulent. But I could likely do that better & more efficiently if I saved the info I needed in a database as opposed to caring at all about the original message.

**Reply by diegovn07** [reply]
Hello, thank you for the response, the situation is that for example: If I send 20 emails of which 5 were not delivered, I will receive 5 delivery status notification failed emails. I need to know which of the 20 emails were the 5 that were not delivered, I am not trying to obtain the emails sent, I am trying to know which email corresponds to each delivery status notification.

The reason for this is that the recipients of the emails are from different domains, some from public institutions and so on, and it is common to receive complaints because they do not receive certain notifications. With this, when someone contacts us, I will be able to verify if the emails are really not delivered and the reason.

**Reply by jstedfast** [reply]
> for example: If I send 20 emails of which 5 were not delivered, I will receive 5 delivery status notification failed emails.

Correct. I'm with you so far...

> I need to know which of the 20 emails were the 5 that were not delivered, I am not trying to obtain the emails sent, I am trying to know which email corresponds to each delivery status notification.

Okay, but aren't those 5 emails "sent"?

I'm not sure that I'm following.

> The reason for this is that the recipients of the emails are from different domains, some from public institutions and so on, and it is common to receive complaints because they do not receive certain notifications. With this, when someone contacts us, I will be able to verify if the emails are really not delivered and the reason.

Are you saying that if one of your customers comes to you asking why a particular message wasn't delivered, you first look up the email that was sent and use that message's Message-Id to look for any Delivery Status Notification failed emails that correspond with that message-id?

**Reply by diegovn07** [reply]
You are right, the emails are not sent, they are not in the sent folder, however from the application I receive an ok, as you said in the other forum "there's no way for smtp.gmail.com to know that user@another-domain. com doesn't exist until it actually attempts to forward the message on to e.g. smtp.another-domain.com."
So yes, I want to obtain the ids of the emails sent to the clients and verify if any had a Delivery Status Notification failed and show something like this in the image
![example](https://github.com/jstedfast/MailKit/assets/66571905/bb7ed291-732c-4689-94bc-77b6905d4997)

**Reply by jstedfast** [reply]
You left me even more confused.
